src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java

```diff
@@ -577,9 +587,62 @@ public void onUpgradeComplete(final int newLevel)
         this.setHeight(wrapper.getHeight());
         this.setCorners(corners.getFirst().getFirst(), corners.getFirst().getSecond(), corners.getSecond().getFirst(), corners.getSecond().getSecond());
         this.isBuilt = true;
+
+        if (newLevel > getBuildingLevel())
+        {
+            final AxisAlignedBB realaabb = getTargetableArea(colony.getWorld());
+            final EntityFireworkRocket firework = new EntityFireworkRocket(colony.getWorld(), realaabb.maxX, realaabb.maxY, realaabb.maxZ, genFireworkItemStack(newLevel));
+
+            colony.getWorld().spawnEntity(firework);
+            final EntityFireworkRocket fireworka = new EntityFireworkRocket(colony.getWorld(), realaabb.maxX, realaabb.maxY, realaabb.minZ, genFireworkItemStack(newLevel));
+
+            colony.getWorld().spawnEntity(fireworka);
+            final EntityFireworkRocket fireworkb = new EntityFireworkRocket(colony.getWorld(), realaabb.minX, realaabb.maxY, realaabb.maxZ, genFireworkItemStack(newLevel));
+
+            colony.getWorld().spawnEntity(fireworkb);
+            final EntityFireworkRocket fireworkc = new EntityFireworkRocket(colony.getWorld(), realaabb.minX, realaabb.maxY, realaabb.minZ, genFireworkItemStack(newLevel));
+
+            colony.getWorld().spawnEntity(fireworkc);
+        }
     }
-    //------------------------- Starting Required Tools/Item handling -------------------------//
 
+    /**
+     * Generates random firework with various properties.
+     *
+     * @return ItemStack of random firework
+     */
+    private ItemStack genFireworkItemStack(final int explosionAmount)
+    {
+        final Random rand = new Random();
+        final ItemStack fireworkItem = new ItemStack(new ItemFirework());
+        final NBTTagCompound itemStackCompound = fireworkItem.getTagCompound() != null ? fireworkItem.getTagCompound() : new NBTTagCompound();
+        final NBTTagCompound fireworksCompound = new NBTTagCompound();
+        final NBTTagList explosionsTagList = new NBTTagList();
+        for (int i = 0; i < explosionAmount; i++)
+        {
+            final NBTTagCompound explosionTag = new NBTTagCompound();
+
+            explosionTag.setBoolean("Flicker", rand.nextInt(2) == 0);
+            explosionTag.setBoolean("Trail", rand.nextInt(2) == 0);
+            explosionTag.setInteger("Type", rand.nextInt(5));
+
+            final int numberOfColours = rand.nextInt(3) + 1;
+            final int[] colors = new int[numberOfColours];
+
+            for (int ia = 0; ia < numberOfColours; ia++)
+            {
+                colors[ia] = ItemDye.DYE_COLORS[rand.nextInt(15)];
+            }
+
+            explosionTag.setIntArray("Colors", colors);
+            explosionsTagList.appendTag(explosionTag);
+        }
+        fireworksCompound.setTag("Explosions", explosionsTagList);
```

Comment:
Extract constants

Reply:
ok I will try to do this

Reply:
btw, you didn't commit your changes if this is resolved =D